[inductor] FX graph cache: Add support for symbolic shapes

masnesral opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #111421

Summary: Add support for caching graphs that have tensor args with symbolic shapes. The high-level appraoch is to serialize guards with the on-disk cached object and validating those guards pass before serving a cached object.

Test Plan: New unit tests

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

masnesral committed:
[RFC][inductor] FX graph cache: Add support for symbolic shapes

Summary: Add support for caching graphs that have tensor args with symbolic shapes. The high-level appraoch is to serialize guards with the on-disk cached object and validating those guards pass before serving a cached object.

Test Plan: New unit tests

[ghstack-poisoned]

masnesral commented on test/inductor/test_codecache.py:
```diff
@@ -152,6 +155,121 @@ def test_cache_load_model(self, device, dtype):
+        self.assertEqual(fn(a, b), compiled_fn(a, b))
+        self.assertEqual(counters["inductor"]["fxgraph_cache_miss"], 1)
+        self.assertEqual(counters["inductor"]["fxgraph_cache_hit"], 1)
+
```
What's missing here is validation that guards were added after loading from the cache. Still investigating, but does anyone know if there's a straightforward way to access the guards in a structured way so I can do some validation here?

masnesral replied:
@ezyang do you have a recommendation? If I have a torch.compiled function, is there a straightforward way to see that our guards were properly added even in the case of a cache hit?

masnesral commented on test/inductor/test_codecache.py:
```diff
@@ -152,6 +155,121 @@ def test_cache_load_model(self, device, dtype):
+            self.assertEqual(fn(x), compiled_fn(x))
+            self.assertEqual(counters["inductor"]["fxgraph_cache_miss"], 1)
+            self.assertEqual(counters["inductor"]["fxgraph_cache_hit"], 1)
+
```
TODO: I also need to check that guards were added properly after a cache hit.

masnesral commented on torch/_inductor/codecache.py:
```diff
@@ -326,16 +327,15 @@ def get_lock_dir() -> str:
-        .decode("utf-8")
-        .lower()
-    )
+    return "c" + sha256_hash(hashing_str)
```
The prefix character seems to be the existing scheme here to differentiate different kinds of hashes for different types of cached objects. Probably I should use an enum. TODO.

masnesral replied:
I was sticking with the convention already started, but can definitely turn it into a a more descriptive string. There's also an existing precedent for adding an extension, e.g., "cg" which I guessed means "code graph"?

masnesral commented on torch/_inductor/codecache.py:
```diff
@@ -553,8 +560,14 @@ class FxGraphHashDetails:
-        self.fx_args = fx_args
+    def __init__(
+        self,
+        gm: torch.fx.GraphModule,
```
Since we need to access the `example_args`, it was convenient to make the `gm` and `example_args` arguments explicit rather than packaging them up in a list.

masnesral commented on torch/_inductor/codecache.py:
```diff
@@ -575,79 +588,193 @@ def __init__(self, fx_args: List[Any], fx_kwargs: Dict[str, Any]) -> None:
         self.inductor_code_hash = get_inductor_code_hash()
 
+    def debug_str(self):
+        """
```
This method is helpful for printing all the fields that we include in the hash when computing the cache key. It is not currently called anywhere, however, because I think it's waaaay too verbose even for debug logging. Does anyone know if we have precedent for different log levels?

eellison replied:
There is log.info vs log.debug. Debug is more verbose

masnesral replied:
Yeah, I'm saying that even log.debug is probably too verbose here. I think you only want to print this level of detail if you need to deeply debug a cache miss that you think shouldn't be a cache miss and you want to see all the gory details of what went into the key calculation.

eellison replied:
You could enable it with an environment variable

ezyang replied:
Don't do an environment variable. Instead, you can create an artifact logger that isn't triggered even in debug. bytecode is done this way, check it out.

masnesral replied:
@ezyang.  Sry, I don't know what bytecode refers to. Is this the sample you have in mind? https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/convert_frame.py#L85

masnesral commented on torch/_inductor/codecache.py:
```diff
@@ -575,79 +588,193 @@ def __init__(self, fx_args: List[Any], fx_kwargs: Dict[str, Any]) -> None:
     # in an in-memory cache after loading from disk.
-    @classmethod
-    def save_graph(cls, key: str, compiled_graph: CompiledFxGraph):
+    @staticmethod
```
I kept this implementation under the `FxGraphCache` class, but all the methods are static. I prefer the namespacing of putting these methods in the class, but lemme know if a class of all static methods hurts your eyeballs.

eellison replied:
Sometimes people prefer `@classmethod` and `cls.method` for this reason i think. up to you.

masnesral commented on torch/_inductor/codecache.py:
```diff
@@ -575,79 +588,193 @@ def __init__(self, fx_args: List[Any], fx_kwargs: Dict[str, Any]) -> None:
+            if hit:
+                # Now re-evaluate to add the guards to the current shape env.
+                # We have to clear the `evaluate_expr` lru_cache to force evaluation.
+                shape_env.evaluate_expr.cache_clear()
```
Please comment. I don't know if clearing the whole cache is problematic, e.g., from a performance perspective. I could, for example, introduce a new context manager analogous to `suppress_guards()` above that selectively uses a cached vs. non-cached version of `shape_env.evaluate_expr`, but I thought it was worth asking if that's overkill.

eellison replied:
Seems off to me that we would clear cache. leaving this to @ezyang to comment

Chillee replied:
I agree it seems kinda weird that we would need to clear the cache. If a cache hit is preventing you from adding a guard, then that should mean that that guard was already evaluated, no?

masnesral replied:
See some of the comments above. The approach here is to:
1. Load a possible match from the cache.
2. Evaluate the guards, but in a mode that _does not_ modify the guards in the current environment, because: in the case that there's a miss, we don't actually want the current env to change.
3. If there's a hit, only then re-evaluate to cause the guards to be added.

The problem here is that the caching at `evaluate_expr()` interferes with #3

ezyang replied:
@masnesral I think a better solution would be to customize the caching on `evaluate_expr` so that it respects whether or not guards were suppressed or not. WDYT?

ezyang replied:
Hmm, well I guess it didn't matter when the hints were ints but it does matter if the hints are symints. I mean, I wouldn't be opposed to just turning off this caching entirely when you're playing funny tricks with symint hints.

masnesral commented on torch/_inductor/codecache.py:
```diff
@@ -659,9 +786,34 @@ class CompiledFxGraph:
 
     _boxed_call: Optional[bool] = None
 
+    def __init__(
```
I preferred a ctor here rather than extracting all these fields at the caller.

masnesral commented on torch/_inductor/compile_fx.py:
```diff
@@ -325,8 +325,7 @@ def compile_fx_inner(
     # Inputs to fx_codegen_and_compile
     # Anything that affects codegen should go here, so if the signature
-    # of fx_codegen_and_compile changes, the list and dict should be updated accordingly
-    graph_args = [gm, example_inputs]
```
See comment above. I changed this to be explicit about the two args (`gm`, `example_inputs`)

masnesral commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -3586,13 +3586,22 @@ def issue_guard(guard: ShapeGuard) -> None:
         self._check_translation_validate()
         return exprs
 
```
The changes here are just about splitting apart the `evaluate_guards_for_args` function into smaller pieces that I can use for the cache impl. Namely, I need separate phases for creating the guards expression and evaluating it (in a different context)

ezyang replied:
Seems like a good comment to put into the code comments

masnesral committed:
fix broken tests on "[RFC][inductor] FX graph cache: Add support for symbolic shapes"


Summary: Add support for caching graphs that have tensor args with symbolic shapes. The high-level appraoch is to serialize guards with the on-disk cached object and validating those guards pass before serving a cached object.

Test Plan: New unit tests

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

masnesral commented on torch/_inductor/codecache.py:
```diff
@@ -575,79 +588,212 @@ def __init__(self, fx_args: List[Any], fx_kwargs: Dict[str, Any]) -> None:
-        write(pickle.dumps(disk_compiled_graph), "cg", extra=key, hash_type="cg")
 
-    @classmethod
-    def load_graph(cls, cg_path: str) -> CompiledFxGraph:
```
The diff looks weird here. `load_graph` is replaced by `lookup_graph` above and the implementation here is for `save_graph`.  So ignore the red lines I guess.

eellison commented on torch/_inductor/codecache.py:
```diff
@@ -491,6 +478,16 @@ def _reduce_tensor(t):
+    """
+    See FxGraphCachePickler. Custom reducer to pickle SymInts.
+    """
+    # For hashing purposes, we only care about the name of the symbol and
```
It might be useful in the future to canonicalize symints so that `hash(tensor([s1, s1, s2])) == hash(tensor([s3, s3, s4]))`.. not needed now

ezyang replied:
It's not clear to me that this is sound, given https://github.com/pytorch/pytorch/pull/111421/files#diff-c9b517f8db609ffa866804dfa2689188a4fee20abacaa0b0dca91625c1b5cb8dR705

if you say [s0, s1] == [s4, s3], you have to make sure that you know how to flip the SymInt arguments when eval'ing the guards. That sounds difficult.

eellison commented on torch/_inductor/codecache.py:
```diff
@@ -575,79 +588,212 @@ def __init__(self, fx_args: List[Any], fx_kwargs: Dict[str, Any]) -> None:
+        # Use a hash of the serialized CompiledFxGraph to get a unique file
+        # name. The specific name doesn't matter since a lookup involves
+        # iterating over all entries in the parent subdir.
+        path = os.path.join(subdir, sha256_hash(content) + ".cg")
+        write_atomic(path, content)
```
do we need FileLock here ?

eellison reviewed (commented):
looks good but let's wait for full review from @ezyang

Chillee commented on torch/_inductor/codecache.py:
```diff
@@ -575,79 +588,212 @@ def __init__(self, fx_args: List[Any], fx_kwargs: Dict[str, Any]) -> None:
+      FxGraph (the graph module, graph inputs, system settings etc.) into an
+      FxGraphCacheDetails object, pickle it, and compute a hash for the key.
+      See FxGraphCachePickler.
+    - Among the metadata we store, we also include a guards expression that's
```
Is it possible to end up loading a more generic graph from cache than we need?

masnesral replied:
Ohhhh, yes. That's a nice catch. So really I need to consider all versions and pick the "best" option, don't I? Hmm, is there any obvious criteria I can use to determine whether one option is more appropriate than the other? cc @eellison

eellison replied:
I think maybe as a follow up - not needed in initial pr imo. Generally the behavior for symbolic shapes is to reuse a compilation if it works.

ezyang replied:
Suppose you have a cache for `(s0, 4)` and another for `(5, s1)`. There is no unique best choice cache entry to load.

Probably the only way to actually do this is some sort of autotuning based thing, where for any given concrete size you've benchmarked which one runs fastest, and you load that particular one.

ani300 commented on test/inductor/test_codecache.py:
```diff
@@ -125,7 +127,8 @@ def fn(x, y):
     @parametrize("device", ("cuda", "cpu"))
     @parametrize("dtype", (torch.float32, torch.bfloat16))
-    def test_cache_load_model(self, device, dtype):
+    @parametrize("dynamic", (False, True))
```
Same as previous comment

ani300 commented on test/inductor/test_codecache.py:
```diff
@@ -152,6 +155,121 @@ def test_cache_load_model(self, device, dtype):
+
+        # Mark all tensor arg dimensions as dynamic to cause all shapes
+        # to be symbolic
+        def rand_dynamic(*args):
```
this is equivalent to dynamic=True

ani300 reviewed (commented):
I made some minor comments on the unit tests wrt the behavior of the `dynamic` parameter. Other than that, not much more to add as @eellison and @Chillee already covered much of the logic questions I had. I defer to @ezyang to see some not-as-obvious issue with the treatment of symbolic shapes, as I'm def not the expert there.

ezyang commented on test/inductor/test_codecache.py:
```diff
@@ -152,6 +155,121 @@ def test_cache_load_model(self, device, dtype):
+        self.assertEqual(counters["inductor"]["fxgraph_cache_hit"], 0)
+
+        # A second call should hit.
+        torch._dynamo.reset()
```
do we have an API for resetting the on disk compiled graph cache?

masnesral replied:
We don't, but I could add one. Do you have a use case in mind where we'd want to wire it up, or you just want to make it available? There are other caches in this file and the clear() methods don't remove on-disk artifacts. Would we want to add those? Would we want to an API to just blow away all artifacts (independent of type)?

ezyang replied:
The testing API probably looks a bit different; instead of blowing away the full cache, you probably just want to cache into some hermetic directory separate from everything else.

ezyang commented on test/inductor/test_codecache.py:
```diff
@@ -152,6 +155,121 @@ def test_cache_load_model(self, device, dtype):
+        self.assertEqual(counters["inductor"]["fxgraph_cache_hit"], 0)
+
+        # A second call should hit.
+        torch._dynamo.reset()
```
are you resetting the counters on each test iteration? It would be easier to understand and refactor the tests if you do

masnesral replied:
Each individual test resets the counters. Is that what you mean? Or you were expecting resets within this individual test body?

ezyang replied:
resets within the test body. the main idea is that if you wobble an early count, you now no longer have to modify all the downstream counts

ezyang commented on test/inductor/test_codecache.py:
```diff
@@ -152,6 +155,121 @@ def test_cache_load_model(self, device, dtype):
+        torch._dynamo.reset()
+        self.assertEqual(fn(a, b), compiled_fn(a, b))
+        self.assertEqual(counters["inductor"]["fxgraph_cache_miss"], 3)
+        self.assertEqual(counters["inductor"]["fxgraph_cache_hit"], 3)
```
So I'm thinking about the test you wrote here, and it seems to me that many pre-existing tests that are checking cache behavior could be repurposed for the FX graph cache. The general idea is that as long as we aren't testing for guards produced by Dynamo, the test can be run in two ways: once without clearing the in memory code cache, and once clearing the in memory code cache, so you hit the FX graph cache. Something to think about.

masnesral replied:
There is no in-memory FX graph cache in the implementation so far though. It's only on disk. Do you mean clearing on the on-disk cache? (btw, I have a TODO to investigate whether extending to an in-memory cache too would help compile time, but my guess is that it will be in the noise)

ezyang replied:
The "in-memory" cache is the Dynamo compiled code cache

ezyang commented on torch/_inductor/codecache.py:
```diff
@@ -326,16 +327,15 @@ def get_lock_dir() -> str:
 
 
+def sha256_hash(data: bytes) -> str:
+    return base64.b32encode(hashlib.sha256(data).digest())[:51].decode("utf-8").lower()
```
Why are we truncating the hash?

Chillee replied:
Probably worth adding a comment haha.

ezyang commented on torch/_inductor/codecache.py:
```diff
@@ -575,79 +588,212 @@ def __init__(self, fx_args: List[Any], fx_kwargs: Dict[str, Any]) -> None:
+        deserialized CompiledFxGraph object. On a miss, return None.
+        """
+        subdir = FxGraphCache._get_subdir(key)
+        if not os.path.exists(subdir):
```
This function needs more logging.

ezyang replied:
In particular, eventually we're going to be debugging weird production issues related to cache related problems, and I need 100% of the relevant file system state we probe to be reported by the logs here, because it's going to be extremely inconvenient to actually get to the machine that has the problem

masnesral replied:
@ezyang it's related to the discussion above about using an env var. I assumed that dumping all this info is going to waaay too verbose for normal debug output -- because it's a lot. Lemme know if that assumption is wrong and I'll just log.debug it. I haven't investigated what you mean by using an artifact logger, but does artifact logging the info give you the control you need?

ezyang replied:
Yes. You can configure artifact logging so it won't get triggered if you do `+inductor`. But I think that there is probably a non-noisy version of the logging which is acceptable on `+inductor` by default. If you are O(1) log lines per inductor compilation that is DEFINITELY acceptable

ezyang commented on torch/_inductor/codecache.py:
```diff
@@ -575,79 +588,212 @@ def __init__(self, fx_args: List[Any], fx_kwargs: Dict[str, Any]) -> None:
+
+        # Iterate over any entries in the subdir for this key and evaluate
+        # their guards to determine whether there's a hit.
+        for path in os.listdir(subdir):
```
Let's sort the listdir so that it is deterministic

ezyang commented on torch/_inductor/codecache.py:
```diff
@@ -575,79 +588,212 @@ def __init__(self, fx_args: List[Any], fx_kwargs: Dict[str, Any]) -> None:
+
+            # Evaluate the guard expression in the current context.
+            shape_env = FxGraphCache._get_shape_env()
+            symints = FxGraphCache._filter_symints(example_inputs)
```
Is this guaranteed to be in the correct order?

ezyang commented on torch/_inductor/codecache.py:
```diff
@@ -575,79 +588,212 @@ def __init__(self, fx_args: List[Any], fx_kwargs: Dict[str, Any]) -> None:
+            # unless we have a hit.
+            with shape_env.suppress_guards():
+                res = shape_env.evaluate_guards_expression(graph.guards_expr, symints)
+                hit = definitely_true(res)
```
You don't need `definitely_true` here, just `bool` it. definitely_true is for dealing with unbacked SymInt

masnesral replied:
thanks!

ezyang commented:
The basic logic looks OK. I'm a little nervous about just slamming `filter_symints` but I think it'll be OK since we are hashing on the FX graph.

masnesral commented on test/inductor/test_codecache.py:
```diff
@@ -85,7 +86,8 @@ def setUp(self):
     @parametrize("device", ("cuda", "cpu"))
     @parametrize("dtype", (torch.float32, torch.bfloat16))
-    def test_cache_load_function(self, device, dtype):
+    @parametrize("dynamic", (False, True))
```
So it turns out there's a shortcoming with that change. None seems to be equivalent to False in terms of the FX graph that gets generated (at least for these tests). So I'd need to rearchitect the tests slightly to make sure each test gets a clean tmp directory. I can do that, but I sorta liked the current behavior of leaving the tmp directory intact for the run of the full set of tests because we'd catch cases of a cache hit when we shouldn't (see my comment on `setUpClass`)

masnesral committed:
reviewer comments on "[RFC][inductor] FX graph cache: Add support for symbolic shapes"


Summary: Add support for caching graphs that have tensor args with symbolic shapes. The high-level appraoch is to serialize guards with the on-disk cached object and validating those guards pass before serving a cached object.

Test Plan: New unit tests

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

masnesral committed:
rebase on "[RFC][inductor] FX graph cache: Add support for symbolic shapes"


Summary: Add support for caching graphs that have tensor args with symbolic shapes. The high-level appraoch is to serialize guards with the on-disk cached object and validating those guards pass before serving a cached object.

Test Plan: New unit tests

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

masnesral committed:
Update on "[RFC][inductor] FX graph cache: Add support for symbolic shapes"


Summary: Add support for caching graphs that have tensor args with symbolic shapes. The high-level appraoch is to serialize guards with the on-disk cached object and validating those guards pass before serving a cached object.

Test Plan: New unit tests

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

masnesral committed:
Update on "[RFC][inductor] FX graph cache: Add support for symbolic shapes"


Summary: Add support for caching graphs that have tensor args with symbolic shapes. The high-level appraoch is to serialize guards with the on-disk cached object and validating those guards pass before serving a cached object.

Test Plan: New unit tests

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

masnesral committed:
Update on "[RFC][inductor] FX graph cache: Add support for symbolic shapes"


Summary: Add support for caching graphs that have tensor args with symbolic shapes. The high-level appraoch is to serialize guards with the on-disk cached object and validating those guards pass before serving a cached object.

Test Plan: New unit tests

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

masnesral committed:
Update on "[RFC][inductor] FX graph cache: Add support for symbolic shapes"


Summary: Add support for caching graphs that have tensor args with symbolic shapes. The high-level appraoch is to serialize guards with the on-disk cached object and validating those guards pass before serving a cached object.

Test Plan: New unit tests

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

masnesral committed:
Update on "[inductor] FX graph cache: Add support for symbolic shapes"


Summary: Add support for caching graphs that have tensor args with symbolic shapes. The high-level appraoch is to serialize guards with the on-disk cached object and validating those guards pass before serving a cached object.

Test Plan: New unit tests

[ghstack-poisoned]

masnesral committed:
rebase on "[inductor] FX graph cache: Add support for symbolic shapes"


Summary: Add support for caching graphs that have tensor args with symbolic shapes. The high-level appraoch is to serialize guards with the on-disk cached object and validating those guards pass before serving a cached object.

Test Plan: New unit tests

[ghstack-poisoned]

masnesral commented:
@pytorchbot merge
